tests/drivers/i2c/i2c_emul/src/test_forwarding_pio.cpp

```diff
@@ -0,0 +1,254 @@
+			.flags = I2C_MSG_WRITE,
+		},
+		{
+			.buf = read_data,
+			.len = sizeof(read_data),
+			.flags = I2C_MSG_READ | I2C_MSG_STOP,
+		},
+	};
+
+	int phase = 0;
+	target_write_requested_0_fake.custom_fake = [&phase](struct i2c_target_config *) -> int {
+		zassert_equal(0, phase, "Expected a call to write_requested before anything else");
+		phase++;
+		return 0;
+	};
+	target_write_received_0_fake.custom_fake = [&phase](struct i2c_target_config *,
+							    uint8_t) -> int {
+		zassert_equal(1, phase, "Expected a call to write_received as the second step");
+		phase++;
+		return 0;
+	};
+	target_read_requested_0_fake.custom_fake = [&phase](struct i2c_target_config *,
+							    uint8_t *) -> int {
+		zassert_equal(2, phase, "Expected a call to read_requested as the 3rd step");
+		phase++;
+		return 0;
+	};
+	target_read_processed_0_fake.custom_fake = [&phase](struct i2c_target_config *,
+							    uint8_t *) -> int {
+		zassert_equal(3, phase, "Expected a call to read_processed as the 4th step");
+		phase++;
+		return 0;
+	};
+	target_stop_0_fake.custom_fake = [&phase](struct i2c_target_config *) -> int {
+		zassert_equal(4, phase, "Expected a call to stop as the 5th step");
+		phase++;
+		return 0;
+	};
+	zassert_ok(i2c_transfer(controller, msgs, ARRAY_SIZE(msgs),
+				emulated_target_config[0].address));
+	zexpect_equal(1, target_write_requested_0_fake.call_count,
+		      "Expected target_write_requested to be called once, but got %d",
+		      target_write_requested_0_fake.call_count);
+	zexpect_equal(1, target_write_received_0_fake.call_count,
+		      "Expected target_write_received to be called once, but got %d",
+		      target_write_received_0_fake.call_count);
+	zexpect_equal(1, target_read_requested_0_fake.call_count,
+		      "Expected target_read_requested to be called once, but got %d",
+		      target_read_requested_0_fake.call_count);
+	zexpect_equal(1, target_read_processed_0_fake.call_count,
+		      "Expected target_read_processed to be called once, but got %d",
+		      target_read_processed_0_fake.call_count);
+	zexpect_equal(1, target_stop_0_fake.call_count,
+		      "Expected target_stop to be called once, but got %d",
+		      target_stop_0_fake.call_count);
+	zexpect_equal(5, phase, "Expected a total of 5 phases, but got %d", phase);
+}
+
+ZTEST(i2c_emul_forwarding, test_forward_two_targets)
+{
+	uint8_t read_data[2];
+
+	// Read the second forward and ensure that we only forwarded to the correct one
+	zassert_ok(i2c_read(controller, read_data, sizeof(read_data),
+			    emulated_target_config[1].address));
+
+	// Check that we got the forward
+	zexpect_equal(1, target_read_requested_1_fake.call_count,
+		      "Expected to be called 1 time, got %d",
+		      target_read_requested_1_fake.call_count);
+	zexpect_equal(1, target_read_processed_1_fake.call_count,
+		      "Expected to be called 1 time, got %d",
+		      target_read_processed_1_fake.call_count);
+
+	// Check that we didn't forward to the first target
```

Comment:
Don't think we allow any comments besides /* */ in our style guide

Reply:
I checked https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#main-rules and rule 3.2 explicitly mentions // comments though I'm not seeing any rules prohibiting them. Are there any specific objections to them in .cpp files? I 100% agree we can't use them in .h files and clearly .c

Reply:
I too couldn't find any rules about // comments after hunting around, though I very much recall a linter/styler checker/our guide saying something along those lines previously its clearly not bothering to check these .cpp files, and there's no definitive rules.